src/D2L.CodeStyle.Analyzers/Pinning/PinnedAttributeAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Pinning {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class PinnedAttributeAnalyzer : DiagnosticAnalyzer {
+
+		private static readonly TypeKind[] TypeKindsSupported = new[] { TypeKind.Class, TypeKind.Interface };
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.PinnedTypesMustNotMove
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( OnCompilationStart );
+		}
+
+		private void OnCompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSymbolAction( AnalyzeSymbol,
+				SymbolKind.NamedType );
+		}
+
+		private static void AnalyzeSymbol( SymbolAnalysisContext context ) {
+			INamedTypeSymbol? pinnedAttributeSymbol = context.Compilation.GetTypeByMetadataName( PinnedAnalyzerHelper.PinnedAttributeName );
+			if( pinnedAttributeSymbol == null ) {
+				return;
+			}
+
+			INamedTypeSymbol? classSymbol = context.Symbol as INamedTypeSymbol;
+			if( classSymbol == null ) {
+				return;
+			}
+
+			if( !TypeKindsSupported.Contains( classSymbol.TypeKind ) ) {
+				return;
+			}
+
+			Location? location = classSymbol.Locations.FirstOrDefault();
+			if( !PinnedAnalyzerHelper.TryGetPinnedAttribute( classSymbol, pinnedAttributeSymbol, out AttributeData? attribute ) ) {
+				return;
+			}
+
+			string? fqName = attribute?.ConstructorArguments[0].Value?.ToString();
+			string? assembly = attribute?.ConstructorArguments[1].Value?.ToString();
+			if( fqName == null || assembly == null ) {

```

Comment:
I think for these to be `null` it would be a compiler error. I guess it doesn't hurt to report the diagnostic, and it's still technically correct